keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/diff/DualNumber.java

```diff
@@ -1,434 +0,0 @@
-        } else {
-            DoubleTensor dArcTan = value.pow(2).plusInPlace(1).reciprocalInPlace();
-            return new DualNumber(newValue, this.partialDerivatives.multiplyAlongOfDimensions(dArcTan, this.getValue().getShape()));
-        }
-    }
-
-    public DualNumber log() {
-        DoubleTensor newValue = value.log();
-        if (this.partialDerivatives.isEmpty()) {
-            return new DualNumber(newValue, PartialDerivatives.OF_CONSTANT);
-        } else {
-            return new DualNumber(newValue, this.partialDerivatives.divideBy(value));
-        }
-    }
-
-    public DualNumber sum() {
-        DoubleTensor sumOfAll = DoubleTensor.scalar(value.sum());
-        int[] resultDims = TensorShape.dimensionRange(0, value.getRank());
-        return new DualNumber(sumOfAll, this.partialDerivatives.sumOverOfDimensions(resultDims));
-    }
-
-    public DualNumber reshape(int[] proposedShape) {
-        PartialDerivatives reshapedPartialDerivatives = this.partialDerivatives.reshape(getValue().getRank(), proposedShape);
-        return new DualNumber(value.reshape(proposedShape), reshapedPartialDerivatives);
-    }
-
-    public DualNumber slice(int dimension, int index) {
-        DoubleTensor newValue = value.slice(dimension, index);
-        boolean needReshape = newValue.getRank() == value.getRank();
-        PartialDerivatives slicedPartialDerivatives = this.partialDerivatives.slice(dimension, index, needReshape);
-        return new DualNumber(newValue, slicedPartialDerivatives);
-    }
-
-    public DualNumber take(int... index) {
-        Map<VertexId, DoubleTensor> dualsAtIndex = new HashMap<>();
-        DoubleTensor newValue = DoubleTensor.scalar(this.value.getValue(index));
-
-        for (Map.Entry<VertexId, DoubleTensor> entry : this.partialDerivatives.asMap().entrySet()) {
-            DoubleTensor atIndexTensor = takeFromPartial(entry.getValue(), index);
-            int desiredRank = atIndexTensor.getShape().length + newValue.getShape().length;
-            int[] paddedShape = TensorShape.shapeToDesiredRankByPrependingOnes(atIndexTensor.getShape(), desiredRank);
-            atIndexTensor = atIndexTensor.reshape(paddedShape);
-            dualsAtIndex.put(entry.getKey(), atIndexTensor);
-        }
-
-        return new DualNumber(newValue, dualsAtIndex);
-    }
-
-    private DoubleTensor takeFromPartial(DoubleTensor from, int... indices) {
-        int[] fromShape = from.getShape();
-        int[] subFromShape = Arrays.copyOf(fromShape, indices.length);
-        int indexToTakeFrom = TensorShape.getFlatIndex(subFromShape, TensorShape.getRowFirstStride(subFromShape), indices);
-        int[] takeShape = Arrays.copyOfRange(fromShape, indices.length, fromShape.length);
-        int subShapeLength = (int) TensorShape.getLength(subFromShape);
-
-        return from.reshape(subShapeLength, -1)
-            .slice(0, indexToTakeFrom)
-            .reshape(takeShape);
-    }
-
-}
```

Comment:
Fun fact: `DualNumber` was our 6th largest source file (in terms of number of lines of code)

    find keanu-project/src/main/ -exec wc -l {} \+ | sort

Reply:
👍 small classes?